File: java/cryptunnel-core/src/main/java/io/github/objectyan/cryptunnel/core/crypto/Sm4Cipher.java
package io.github.objectyan.cryptunnel.core.crypto;

import org.bouncycastle.jce.provider.BouncyCastleProvider;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.Provider;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;
import java.util.Collection;
import java.util.Collections;

/**
 * 国密 SM4 隧道加密算法：SM4-CBC + HMAC-SHA256（encrypt-then-MAC）。
 *
 * <p>JDK 标准 Provider（SunJCE）在任何版本都不提供 SM4，因此本实现依赖 BouncyCastle。
 * BouncyCastle 在 core 中是 <b>optional 依赖</b>：不使用 SM4 时无需引入，
 * 也不会传递给宿主应用。</p>
 *
 * <p>载荷格式与默认的 {@link AesCbcHmacSha256Cipher} <b>完全一致</b>，
 * 只是把分组算法从 AES 换成 SM4——这样三端改造只需替换 Cipher 算法名，
 * 帧结构、HMAC 覆盖范围、IV 长度全部不变，跨语言对齐成本最低：</p>
 * <pre>
 *   Base64( IV[16] || HMAC[32] || SM4/CBC/PKCS7Padding 密文 )
 *   HMAC 覆盖范围：IV || 密文
 *   密钥派生：SHA-256("SM4:" + rawKey) 取前 16 字节做 SM4-128 密钥
 *            HMAC 密钥仍为 SHA-256("HMAC:" + rawKey)，与默认算法一致
 * </pre>
 *
 * <p>采用独立派生前缀 {@code "SM4:"}，使同一把 rawKey 下的 SM4 密钥与 AES 密钥
 * 互不相关（避免算法间密钥复用带来的推导风险）。</p>
 *
 * <p><b>Provider 使用方式</b>：显式传入 {@link Provider} 实例而非调用
 * {@code Security.addProvider(...)}，避免污染宿主应用的全局 JCE 配置。</p>
 *
 * <p><b>可用性</b>：classpath 上不存在 BouncyCastle 时，{@link TunnelCiphers} 不会注册本实现，
 * 配置 {@code cipher: sm4} 会给出「缺少依赖」的明确报错，而不是 NoClassDefFoundError。</p>
 */
public final class Sm4Cipher implements TunnelCipher {

    /** 算法标识，用于配置匹配；不进报文。 */
    public static final String ID = "sm4-cbc-hmac-sha256";

    /** 简便别名，yml 里写 {@code cipher: sm4} 亦可。 */
    public static final String ALIAS = "sm4";

    private static final String HMAC_ALGORITHM = "HmacSHA256";

    private static final int IV_LENGTH = 16;
    private static final int HMAC_LENGTH = 32;
    private static final int SM4_KEY_LENGTH = 16;

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    /** 显式 Provider 实例，避免全局注册影响宿主应用。 */
    private static final Provider BC = new BouncyCastleProvider();

    /** BC 的 SM4 变换名：新版本用 PKCS7Padding，旧版本只有 PKCS5Padding，故运行时探测。 */
    private static final String TRANSFORMATION = resolveTransformation();

    @Override
    public String id() {
        return ID;
    }

    @Override
    public Collection<String> aliases() {
        return Collections.singletonList(ALIAS);
    }

    @Override
    public String seal(byte[] plaintext, String rawKey) {
        try {
            byte[] iv = new byte[IV_LENGTH];
            SECURE_RANDOM.nextBytes(iv);
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            Cipher cipher = Cipher.getInstance(TRANSFORMATION, BC);
            cipher.init(Cipher.ENCRYPT_MODE, deriveSm4Key(rawKey), ivSpec);
            byte[] ciphertext = cipher.doFinal(plaintext);

            Mac mac = Mac.getInstance(HMAC_ALGORITHM);
            mac.init(AesUtil.deriveHmacKey(rawKey));
            byte[] hmacBytes = mac.doFinal(AesUtil.concat(iv, ciphertext));

            return Base64.getEncoder().encodeToString(AesUtil.concat(iv, hmacBytes, ciphertext));
        } catch (Exception e) {
            throw new TunnelCryptoException("sm4 seal failed: " + e.getMessage(), e);
        }
    }

    @Override
    public byte[] open(String payload, String rawKey) {
        try {
            byte[] raw = Base64.getDecoder().decode(payload);
            if (raw.length < IV_LENGTH + HMAC_LENGTH) {
                throw new TunnelCryptoException("sm4 payload too short: " + raw.length + " bytes");
            }

            byte[] iv = AesUtil.subarray(raw, 0, IV_LENGTH);
            byte[] hmacReceived = AesUtil.subarray(raw, IV_LENGTH, IV_LENGTH + HMAC_LENGTH);
            byte[] ciphertext = AesUtil.subarray(raw, IV_LENGTH + HMAC_LENGTH, raw.length);

            Mac mac = Mac.getInstance(HMAC_ALGORITHM);
            mac.init(AesUtil.deriveHmacKey(rawKey));
            byte[] hmacComputed = mac.doFinal(AesUtil.concat(iv, ciphertext));
            if (!MessageDigest.isEqual(hmacReceived, hmacComputed)) {
                throw new SecurityException("HMAC verification failed - data may be tampered");
            }

            Cipher cipher = Cipher.getInstance(TRANSFORMATION, BC);
            cipher.init(Cipher.DECRYPT_MODE, deriveSm4Key(rawKey), new IvParameterSpec(iv));
            return cipher.doFinal(ciphertext);
        } catch (RuntimeException e) {
            throw new TunnelCryptoException("sm4 open failed: " + e.getMessage(), e);
        } catch (Exception e) {
            throw new TunnelCryptoException("sm4 open failed: " + e.getMessage(), e);
        }
    }

    /**
     * @return SM4-128 密钥：SHA-256("SM4:" + rawKey) 的前 16 字节
     */
    private static SecretKeySpec deriveSm4Key(String rawKey) {
        try {
            MessageDigest sha = MessageDigest.getInstance("SHA-256");
            byte[] digest = sha.digest(("SM4:" + rawKey).getBytes(StandardCharsets.UTF_8));
            return new SecretKeySpec(Arrays.copyOf(digest, SM4_KEY_LENGTH), "SM4");
        } catch (Exception e) {
            throw new TunnelCryptoException("failed to derive SM4 key", e);
        }
    }

    private static String resolveTransformation() {
        String[] candidates = {"SM4/CBC/PKCS7Padding", "SM4/CBC/PKCS5Padding"};
        for (String t : candidates) {
            try {
                Cipher.getInstance(t, BC);
                return t;
            } catch (Exception ignored) {
                // 试下一个
            }
        }
        throw new TunnelCryptoException(
                "BouncyCastle 未提供可用的 SM4/CBC 变换，请升级 bcprov 版本");
    }
}

File: dotnet/src/Cryptunnel.Core/Crypto/Sm4Cipher.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace Cryptunnel.Core.Crypto;

/// <summary>
/// 国密 SM4 隧道加密：SM4-CBC + HMAC-SHA256（encrypt-then-MAC），
/// 与 Java 侧 <c>Sm4Cipher</c> 字节级对应。
///
/// 载荷格式与默认算法 <see cref="AesCbcHmacSha256Cipher"/> <b>完全一致</b>，
/// 只把分组算法从 AES 换成 SM4——帧结构、HMAC 覆盖范围、IV 长度全都不变：
/// <code>
///   Base64( IV[16] || HMAC-SHA256[32] || SM4/CBC/PKCS7 密文 )
///   HMAC 覆盖范围：IV || 密文（不含 HMAC 字段本身）
///   SM4  密钥：SHA256("SM4:"  + rawKey) 的前 16 字节
///   HMAC 密钥：SHA256("HMAC:" + rawKey)（与默认算法共用）
/// </code>
///
/// <b>关于 PKCS7 与 Java 的 PKCS5Padding</b>：Java/BC 对 16 字节分组算法所谓的
/// "PKCS5Padding" 实际就是 PKCS7，两端填充逐字节相同。
///
/// <b>为什么不用 BouncyCastle</b>：见 <see cref="Sm4Engine"/> 的说明——
/// 分组函数自研 + 官方测试向量守护，避免给单文件客户端加 4MB 第三方依赖。
/// </summary>
public sealed class Sm4Cipher : ITunnelCipher
{
    /// <summary>算法标识，与 Java 侧 <c>Sm4Cipher.ID</c> 完全一致。</summary>
    public const string CipherId = "sm4-cbc-hmac-sha256";

    /// <summary>简便别名，配置里写 <c>sm4</c> 亦可。</summary>
    public const string Alias = "sm4";

    private const int IvLength = 16;
    private const int HmacLength = 32;
    private const int HeaderLength = IvLength + HmacLength;
    private const int Sm4KeyLength = 16;

    public string Id => CipherId;

    /// <summary>SM4-128 密钥：SHA256("SM4:" + rawKey) 的前 16 字节。</summary>
    internal static byte[] DeriveSm4Key(string rawKey)
    {
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes("SM4:" + rawKey));
        return digest[..Sm4KeyLength];
    }

    public string Seal(ReadOnlySpan<byte> plaintext, string rawKey)
    {
        var iv = RandomNumberGenerator.GetBytes(IvLength);
        var ciphertext = EncryptCbcPkcs7(plaintext, DeriveSm4Key(rawKey), iv);

        // HMAC 覆盖 (IV || 密文)
        var macInput = new byte[IvLength + ciphertext.Length];
        iv.CopyTo(macInput, 0);
        ciphertext.CopyTo(macInput, IvLength);
        var mac = HMACSHA256.HashData(AesCrypto.DeriveHmacKey(rawKey), macInput);

        var result = new byte[HeaderLength + ciphertext.Length];
        iv.CopyTo(result, 0);
        mac.CopyTo(result, IvLength);
        ciphertext.CopyTo(result, HeaderLength);

        return Convert.ToBase64String(result);
    }

    public byte[] Open(string payload, string rawKey)
    {
        var raw = Convert.FromBase64String(payload);
        if (raw.Length < HeaderLength)
            throw new CryptographicException(
                $"sm4 载荷长度不足：{raw.Length} 字节，至少需要 {HeaderLength} 字节（IV+HMAC）");

        var iv = raw.AsSpan(0, IvLength);
        var receivedMac = raw.AsSpan(IvLength, HmacLength);
        var ciphertext = raw.AsSpan(HeaderLength);

        var macInput = new byte[IvLength + ciphertext.Length];
        iv.CopyTo(macInput);
        ciphertext.CopyTo(macInput.AsSpan(IvLength));
        var computedMac = HMACSHA256.HashData(AesCrypto.DeriveHmacKey(rawKey), macInput);

        if (!CryptographicOperations.FixedTimeEquals(receivedMac, computedMac))
            throw new CryptographicException("HMAC 校验失败，数据可能被篡改或密钥不一致");

        return DecryptCbcPkcs7(ciphertext, DeriveSm4Key(rawKey), iv);
    }

    /// <summary>SM4-CBC 加密 + PKCS7 填充。明文长度为分组整数倍时也补满一整块。</summary>
    private static byte[] EncryptCbcPkcs7(ReadOnlySpan<byte> plaintext, byte[] key, ReadOnlySpan<byte> iv)
    {
        var padLength = Sm4Engine.BlockSize - plaintext.Length % Sm4Engine.BlockSize;
        var total = plaintext.Length + padLength;

        var buffer = new byte[total];
        plaintext.CopyTo(buffer);
        buffer.AsSpan(plaintext.Length).Fill((byte)padLength);

        var engine = new Sm4Engine(key);
        var output = new byte[total];
        Span<byte> previous = stackalloc byte[Sm4Engine.BlockSize];
        iv.CopyTo(previous);
        Span<byte> block = stackalloc byte[Sm4Engine.BlockSize];

        for (var offset = 0; offset < total; offset += Sm4Engine.BlockSize)
        {
            for (var i = 0; i < Sm4Engine.BlockSize; i++)
                block[i] = (byte)(buffer[offset + i] ^ previous[i]);

            var target = output.AsSpan(offset, Sm4Engine.BlockSize);
            engine.EncryptBlock(block, target);
            target.CopyTo(previous);
        }

        CryptographicOperations.ZeroMemory(buffer);
        return output;
    }

    /// <summary>SM4-CBC 解密 + 剥离 PKCS7 填充。</summary>
    private static byte[] DecryptCbcPkcs7(ReadOnlySpan<byte> ciphertext, byte[] key, ReadOnlySpan<byte> iv)
    {
        if (ciphertext.Length == 0 || ciphertext.Length % Sm4Engine.BlockSize != 0)
            throw new CryptographicException(
                $"sm4 密文长度 {ciphertext.Length} 不是分组长度 {Sm4Engine.BlockSize} 的整数倍");

        var engine = new Sm4Engine(key);
        var plain = new byte[ciphertext.Length];
        Span<byte> previous = stackalloc byte[Sm4Engine.BlockSize];
        iv.CopyTo(previous);
        Span<byte> decrypted = stackalloc byte[Sm4Engine.BlockSize];

        for (var offset = 0; offset < ciphertext.Length; offset += Sm4Engine.BlockSize)
        {
            var current = ciphertext.Slice(offset, Sm4Engine.BlockSize);
            engine.DecryptBlock(current, decrypted);
            for (var i = 0; i < Sm4Engine.BlockSize; i++)
                plain[offset + i] = (byte)(decrypted[i] ^ previous[i]);
            current.CopyTo(previous);
        }

        var padLength = plain[^1];
        if (padLength is 0 || padLength > Sm4Engine.BlockSize || padLength > plain.Length)
            throw new CryptographicException("sm4 PKCS7 填充非法，密钥不匹配或数据损坏");
        for (var i = plain.Length - padLength; i < plain.Length; i++)
        {
            if (plain[i] != padLength)
                throw new CryptographicException("sm4 PKCS7 填充非法，密钥不匹配或数据损坏");
        }

        var result = plain[..^padLength];
        CryptographicOperations.ZeroMemory(plain);
        return result;
    }
}
